### PR TITLE
Make MultiplexerClient implement TagCapable

### LINKED ISSUE #338: multiplex client should implement TagInterface
the multiplexer client should also support multiplexing the tag invalidation methods provided by TagInterface

### PR DESCRIPTION
**Pull request by ddeboer**
Fix #338.

### CODE REVIEW

File: src/ProxyClient/MultiplexerClient.php [hunk 1]
```diff
@@ -131,13 +165,19 @@ public function refresh($url, array $headers = [])
...
+        $result = null;
+
         foreach ($this->proxyClients as $proxyClient) {
             if (is_subclass_of($proxyClient, $interface)) {
-                call_user_func_array([$proxyClient, $method], $arguments);
+                $result = call_user_func_array([$proxyClient, $method], $arguments);
             }
         }
+
+        return $result;
```
Comment (ddeboer) [inline src/ProxyClient/MultiplexerClient.php]: We have to return a string for header name and value (not an array), so I’m simply returning the value of the last TagCapable client. Does that make sense? Or should we just return null or some constant string `multiple`?

Comment (dbu) [inline src/ProxyClient/MultiplexerClient.php]: i would hope that usually the clients are the same implementation, so this solution seems fine to me. we should add some note in the documentation of the multipart client about this though.

however, i propose that for the getters we call the first client instead of using invoke. if we agree on that, the changes to the invoke method can be reverted.

File: src/ProxyClient/MultiplexerClient.php [hunk 2]
```diff
@@ -96,6 +98,38 @@ public function flush()
     }
 
     /**
+     * {@inheritdoc}
+     */
+    public function getTagsHeaderValue(array $tags)
+    {
+        return $this->invoke(TagCapable::class, 'getTagsHeaderValue', [$tags]);
```
Comment (dbu) [inline src/ProxyClient/MultiplexerClient.php]: this is not supposed to change anything on the client, so i think we should instead just call the first client and return that, instead of calling all of them and then return the last.

File: src/ProxyClient/MultiplexerClient.php [hunk 3]
```diff
@@ -96,6 +98,38 @@ public function flush()
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getTagsHeaderName()
+    {
+        return $this->invoke(TagCapable::class, 'getTagsHeaderName', []);
```
Comment (dbu) [inline src/ProxyClient/MultiplexerClient.php]: same for this

**Comment by ddeboer** [comment]
Updated. Decided to not add the implementation details to the docs as it would clutter them up too much in my opinion.

File: src/ProxyClient/MultiplexerClient.php [hunk 4]
```diff
@@ -126,18 +160,51 @@ public function refresh($url, array $headers = [])
...
+     * @param string $method    The method to invoke
+     * @param array  $arguments The arguments to be passed to the method
+     *
+     * @return mixed Return value of ProxyClient method call
+     */
+    private function invokeFirst($interface, $method, array $arguments)
+    {
+        foreach ($this->getProxyClients($interface) as $proxyClient) {
+            return call_user_func_array([$proxyClient, $method], $arguments);
+        }
```
Comment (dbu) [inline src/ProxyClient/MultiplexerClient.php]: what should we do when none of the clients implemented the requested interface? a bit hard to say - doing tags will not actually do anything, so we could throw an exception. on the other hand, we might just return a header name and an empty header value in that case, to not break users that check on the interface to know if they can use tags.

or we could create a TagCapableMultiplexerClient extends MultiplexerClient that will fail if none of its wrapped clients is TagCapable

Comment (ddeboer) [inline src/ProxyClient/MultiplexerClient.php]: I’m not sure really. Throwing an exception would be the most explicit way of handling that situation.

Comment (dbu) [inline src/ProxyClient/MultiplexerClient.php]: i guess the multiplexer client has this problem for all operations. it hides what the actual clients support or don't support. so yes, lets throw an exception. i will check if the documentation is explaining the special situation with multiplexer client

Comment (dbu) [inline src/ProxyClient/MultiplexerClient.php]: maybe you can review and merge #346 and then add a bit about exceptions for tag operations in this PR?

File: src/ProxyClient/MultiplexerClient.php [hunk 5]
```diff
@@ -126,18 +161,53 @@ public function refresh($url, array $headers = [])
...
+
+        throw UnsupportedProxyOperationException::cacheDoesNotImplement($interface);
```
Comment (ddeboer) [inline src/ProxyClient/MultiplexerClient.php]: Now throwing exceptions (for `invokeFirst` only). Just being lazy and passing the interface’s FQDN in there, but the exception message is clear enough.

**Comment by dbu** [comment]
thanks, i would say this is definitely good enough for something that is a special solution in the first place. i have restarted the php 5.5 build, it looked more like we have some test hoping to not run over the boundary of a second...

can you merge #346 and then rebase on master to add to the note about the multiplexer client to say that while other capabilities can be silently ignored, the tag operations will throw an exception as they would have to return a value?

**Comment by dbu** [comment]
are we ready for 2.0 with this? or is there anything left to do in the library? we tagged alpha1, so i would say after merging this lets tag beta1? we should then try to make the bundle ready-ish in case that reveals issues with the library.

**Comment by ddeboer** [comment]
Updated docs.

**Comment by dbu** [comment]
great, thanks! i just looked over the bundle - there is quite some stuff in the 2.0 milestone. maybe we should go over that again and see what are quick-wins that we can do easily, and what are must haves, and then kick out everything else to not delay too much. new features can also come in 2.1.
